README.md
# ChAT Model

[![Linux](https://github.com/trfore/chatmodel/actions/workflows/build-linux64.yml/badge.svg?branch=main)](https://github.com/trfore/chatmodel/actions/workflows/build-linux64.yml)
[![MacOS & Win x64](https://github.com/trfore/chatmodel/actions/workflows/build-osx64-win64.yml/badge.svg?branch=main)](https://github.com/trfore/chatmodel/actions/workflows/build-osx64-win64.yml)

## Introduction

This is an implementation of the spiking cerebellar granular layer model described in:

> Taylor R Fore et al. "Acetylcholine modulates cerebellar granule cell spiking by regulating the balance of synaptic excitation and inhibition"
> Journal of Neuroscience 1 April 2020, DOI: https://doi.org/10.1523/JNEUROSCI.2148-19.2020

### Code repository

The model is implemented using the [Brian2 simulator](http://briansimulator.org/) for spiking neural networks, which is written in Python. The two scripts for running the simulation and generating the figures in the paper are:

- **run_simulation.py** - main script to run single instantiation of the model.
- **figures.py** - data analysis and visualization.

## How to Use

### Requirements

```
brian2==2.2.1
matplotlib
numpy>=1.19
pandas
python>=3.6.8,<3.8.0
scipy
scikit-learn
```

You can install these dependencies using either `conda-lock.yml` (recommended), `environment.yml` or `requirements.txt` file.

1. Install [miniconda] (recommended), [conda], or [mamba]
   - We highly recommend changing conda's default solver to `libmamba`, see: https://www.anaconda.com/blog/conda-is-fast-now
2. Clone the repo and `cd` into the directory.
3. Install the dependencies into virtual environment, i.e. `brian2`

```sh
# clone the repo
$ git clone https://github.com/trfore/chatmodel.git
$ cd chatmodel

# Recommended, create 'brian2' environment
$ conda install conda-lock
$ conda-lock install --name brian2 lock_files/conda-lock.yml

# create env using environment.yml
$ conda env create -f environment.yml
# create env using requirements.txt
$ conda create --name brian2 --file requirements.txt --channel conda-forge
```

We recommend using the `conda-lock.yml` file, as it will install validated package versions tested against the following platforms.

### Tested Platforms & Simulation Run Times

- Linux (Debian 10+, Ubuntu 20.04/22.04)
- macOS (10.13, 11, 12)
- Windows (10)

#### Simulation Run Times and Resource Utilization

- The full simulation takes ~ 42 minutes (2020 AMD 5600x, 32 GB) to 5 hours (2013 Intel I5-4258U, 8 GB) to run. This time is dependent on your system hardware.
- Unconstrained, the model uses a maximum of `27 GB` of RAM and `2` physical cores, but will run on systems with less RAM as the simulation will write to `cache` and `swap` at the cost of a longer run time.
- The output files, `*.npy`, are collectively `328 MB` in size. An archive is available on the release page (link: [v1.1.0](https://github.com/trfore/chatmodel/releases/tag/v1.1.0)).

### Running the scripts

Clone the repo and `cd` into the directory.

```sh
$ git clone https://github.com/trfore/chatmodel.git
$ cd chatmodel
```

All three drug conditions (control, GABAzine, muscarine) are built into the main script. To run the full simulation:

```sh
$ python run_simulation.py
```

The code saves all relevant variables as numpy arrays to the current directory for further analysis. Robust progress reporting is output to the console, but expect the entire simulation to take approximately 1 to 3 hours (depending on computer specs, see 'Tested Platforms & Simulation Run Times' for details).

To run analysis and generate the figures from the paper, type in terminal:

```sh
$ python figures.py
```

Figure windows will not appear, but pdf files of each figure will be saved in the working directory.

## Environment Variables

You can change experiment parameters by setting environment variables, for example:

```sh
# Run the simulation for 50 trials
$ export CHAT_NUM_TRIAL=50
$ python run_simulation.py
```

| Environment Variable   | Default | Description                                               | Required |
| ---------------------- | ------- | --------------------------------------------------------- | -------- |
| CHAT_NUM_MF            | 315     | int, number of mossy fiber inputs                         | No       |
| CHAT_NUM_GRC           | 4096    | int, number of granule cells                              | No       |
| CHAT_NUM_GOC           | 27      | int, number of golgi cells                                | No       |
| CHAT_NUM_TRIAL         | 100     | int, number of trials for the simulation                  | No       |
| CHAT_RANDOM_WEIGHTS    | True    | boolean, use random weights for synaptic connections      | No       |
| CHAT_SEED_TRIAL        | 451     | int, seed value for trial runs                            | No       |
| CHAT_SEED_WEIGHT_MEANS | 35      | int, seed value for generating the random synaptic weight | No       |

## Additional Notes

The paper was created using the following packages:

```sh
brian2==2.2.1
numpy==1.15.4
scipy==1.2.0
pandas==0.24.1
matplotlib==3.0.2
scikit-learn=0.19.2
```

The original `requirements.txt` file is available here: [e69794d](https://github.com/trfore/chatmodel/commit/e69794d5f8bfb676a317b3f1624e47d9baaaad4e)

---

# Authors

- Nathan Taylor (https://github.com/taylorbn)
- Taylor Fore (https://github.com/trfore)

## Maintainers

- Taylor Fore

---

# References

- https://briansimulator.org/
- https://github.com/brian-team/brian2
- https://github.com/trfore/chatmodel

## Anaconda & Mamba

- https://www.anaconda.com/download/
- https://docs.conda.io/en/latest/miniconda.html
- https://mamba.readthedocs.io/en/latest/installation.html
- https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html

[conda]: https://www.anaconda.com/download/
[miniconda]: https://docs.conda.io/en/latest/miniconda.html
[mamba]: https://mamba.readthedocs.io/en/latest/installation.html


CHANGELOG.md
# Changelog

## [v1.1.0](https://github.com/trfore/chatmodel/releases/tag/v1.1.0) (2023-05-05)

### Updates:

- feature: include Windows lock file ([de067f9](https://github.com/trfore/chatmodel/commit/de067f9950119fa63e1688f4b7985b0eee0ed0e1))
- feature: accept env vars as input resolves #2 ([3368369](https://github.com/trfore/chatmodel/commit/3368369d70f93c12810bab43fce5c82d8b60d0d0))

### Fixes:

- fix: bug in figures.py x-axis value error #6 ([32a7b91](https://github.com/trfore/chatmodel/commit/32a7b9181643e88efc3cf980047b9e7f49736f37))
- fix: issue #5 that prevented figures.py from using *.npy model files ([d1ebfa0](https://github.com/trfore/chatmodel/commit/d1ebfa0ba3f70feef19abbdb4400dbeec8facff9))
- fix: deprecated functions in figures.py #4 ([5eac384](https://github.com/trfore/chatmodel/commit/5eac384559b09b9511c947c9449a4197ee25e79b))
- fix: deprecated functions in run_simulation.py #3 ([cf669df](https://github.com/trfore/chatmodel/commit/cf669df29c6da8d1af0bbc00ba79bdcb500d9c57))

### Chores:

- chore: update dependencies, numpy>=1.19 ([f32d146](https://github.com/trfore/chatmodel/commit/f32d1463328b2c95e9fcc3d5d8bbfbe84d306261))

---

## [v1.0.1](https://github.com/trfore/chatmodel/releases/tag/v1.0.1) (2023-05-04)

### Notes:

- This release is at parity with the initial release ([v1.0](https://github.com/trfore/chatmodel/releases/tag/v1.0)). It adds additional dependency files for end-user convenience.

### Updates:

- update: move to semver x.x.x

### Chores:

- chore: update dependency files, add lock files ([310476b](https://github.com/trfore/chatmodel/commit/310476bb87bed1d61086ae7f5ee0727c40a68ed0))
- docs: add license ([66807e6](https://github.com/trfore/chatmodel/commit/66807e646717ea1936f0c32911e85cb94cf90c69))

---

## [v1.0](https://github.com/trfore/chatmodel/releases/tag/v1.0) (2021-07-07)

- Initial release for the paper
- https://doi.org/10.1523/JNEUROSCI.2148-19.2020

---


.github/test/test_values.sh
#!/usr/bin/env bash
#
# Description: Quick values that approximate a 25um^3 area, these values are for
# testing only, non-biological. The goal is a simulation that runs within the 
# hardware restrictions of GitHub runners, 7 GB and a job timeout of 6 hrs. 
#
# The paper utilizes cell numbers that approximate a 100um^3 area, which is 
# grounded in anatomical and functional studies of the cerebellum.
export CHAT_NUM_TRIAL=10 CHAT_NUM_MF=78 CHAT_NUM_GRC=1024 CHAT_NUM_GOC=7 CHAT_SEED_TRIAL=451

plotSpikeProb.py
import numpy as np
import matplotlib.pyplot as plt
from collections import Counter

random_weights = True

# Import matrices of individual cell spike probabilities saved for each
# experimental condition
if random_weights:
    control_all = np.load('randomized_controlSpikeProb.npy')
    gaba_all    = np.load('randomized_gabaSpikeProb.npy')
    ach_all     = np.load('randomized_achSpikeProb.npy')
else:
    control_all = np.load('controlSpikeProb.npy')
    gaba_all    = np.load('gabaSpikeProb.npy')
    ach_all     = np.load('achSpikeProb.npy')

# Get rid of entirely quiescent cells
indices_nonzero = ~np.all(control_all==0,axis=1)

control = control_all[indices_nonzero,1000:1060]
gaba = gaba_all[indices_nonzero,1000:1060]
ach = ach_all[indices_nonzero,1000:1060]

ncells = control.shape[0]
nt     = control.shape[1]

# Normalize to control condition (max-min)
controlmax = np.reshape(np.amax(control,axis=1),(ncells,1))

controlnorm = control/controlmax
gabanorm    = gaba/controlmax
achnorm     = ach/controlmax

# Average
max_ach = np.amax(achnorm,axis=1)
ach_increase = max_ach > 1

mean_control_up = np.mean(controlnorm[ach_increase,:],axis=0)
mean_gaba_up    = np.mean(gabanorm[ach_increase,:],axis=0)
mean_ach_up     = np.mean(achnorm[ach_increase,:],axis=0)

mean_control_down = np.mean(controlnorm[~ach_increase,:],axis=0)
mean_gaba_down    = np.mean(gabanorm[~ach_increase,:],axis=0)
mean_ach_down     = np.mean(achnorm[~ach_increase,:],axis=0)

stim_times = [1010, 1020, 1030]

fig, ax = plt.subplots(1, 3, figsize=(12,4))
ax[0].plot(np.arange(nt),mean_control_up,'k')
ax[0].plot(np.arange(nt),mean_gaba_up,'r')
ax[0].plot(np.arange(nt),mean_ach_up,'b')


ax[1].plot(np.arange(nt),mean_control_down,'k')
ax[1].plot(np.arange(nt),mean_gaba_down,'r')
ax[1].plot(np.arange(nt),mean_ach_down,'b')

x = np.amax(control,axis=1)
y = np.amax(ach, axis=1)
c = Counter(zip(x,y))
s = [2*c[(xx,yy)] for xx,yy in zip(x,y)]
ax[2].scatter(x,y,c='k',s=s)
ax[2].set_xlabel('Control peak spike probability')
ax[2].set_ylabel('Muscarine peak spike probability')
start,stop = ax[2].get_xlim()
x = np.linspace(start,stop)
y = x
ax[2].plot(x,y,'--k')
plt.show(block=False)


run_simulation.py
from brian2 import *
from matplotlib import *
from sympy import *
import numpy as np
import pandas as pd
import scipy.sparse
import random
import os

'''
Environment Variables
---------------------
CHAT_NUM_MF : int, default = 315
    number of mossy fiber inputs
CHAT_NUM_GRC : int, default = 4096
    number of granule cells
CHAT_NUM_GOC : int, default = 27
    number of golgi cells
CHAT_NUM_TRIAL : int, default = 100
    number of trials for the simulation
CHAT_RANDOM_WEIGHTS : boolean, default = True
    use random weights for synaptic connections
CHAT_SEED_TRIAL : int, default = 451
    seed value for trial runs
CHAT_SEED_WEIGHT_MEANS : int, default = 35
    seed value for generating the random synaptic weight 

Use:
$ export CHAT_NUM_TRIAL=50
$ export CHAT_NUM_TRIAL=100 CHAT_NUM_MF=315 CHAT_NUM_GRC=4096 CHAT_NUM_GOC=27 # 100um^3 (published: biologically relevant)
'''

def main():
    Conditions = ['Control','GABAzine','ACh']
    for Condition in Conditions:
        print('Running {} condition'.format(Condition))
        trial_sim(Condition)
    print('done')

def trial_sim(Condition):

    @check_units(mu=1,sigma=1,result=1)
    def gamma_parms(mu,sigma):
        a,theta = symbols('a theta')
        eqns = [Eq(a*theta,mu), Eq(a*theta**2,sigma**2)]
        vars = [a,theta]
        sols = nonlinsolve(eqns,vars)
        sol_a,sol_theta = next(iter(sols))
        shape = float(sol_a)
        scale = float(sol_theta)
        return shape, scale

    seed(int(os.getenv("CHAT_SEED_TRIAL", 451)))
    ##############
    # Parameters #
    ##############

    # Trial parameters
    runtime = 2000

    # Synaptic weight parameters
    random_weights = bool(os.getenv("CHAT_RANDOM_WEIGHTS", True))
    # weights_gamma_shape,weights_gamma_scale = gamma_parms(0.15,0.2)
    weights_gamma_shape,weights_gamma_scale = gamma_parms(0.1,0.1)

    # Cell numbers to reproduce functionally relevant 100um^3 cube of granular layer
    nmf  = int(os.getenv("CHAT_NUM_MF", 315))   # Mossy fibers
    nGrC = int(os.getenv("CHAT_NUM_GRC", 4096)) # Granule cells
    nGoC = int(os.getenv("CHAT_NUM_GOC", 27))   # Golgi cells

    # Convergence ratios of connections (to determine connection probabilities)
    conv_GrC_M   =   4          #  mf -> GrC synapses
    conv_GoC_M   =  10          #  mf -> GoC synapses
    conv_GoC_GrC = 100          # GrC -> GoC synapses
    conv_GrC_GoC =   4          # Goc -> GrC synapses

    # Leak conductance
    g_l_GrC   = 0.2   * nS
    g_t_GrC   = 1.0   * nS
    g_l_GoC   = 3.    * nS

    # Reversal potential (leak, excitatory, inhibitory)
    E_l_GrC   = -75   * mV
    E_e_GrC   =   0   * mV
    E_i_GrC   = -75   * mV

    E_l_GoC   = -51   * mV
    E_e_GoC   =   0   * mV
    E_i_GoC   = -75   * mV

    # Membrane capacitance
    C_m_GrC   =  3.1  * pF
    C_m_GoC   = 60.   * pF

    # Decay constants
    tau_e_decay_GrC = 12.0 * ms
    tau_e_decay_GoC = 12.0 * ms

    tau_i_decay_GrC = 20.0 * ms

    # Absolute refractory period
    tau_r_GrC = 2 * ms
    tau_r_GoC = 10* ms

    # Spiking threshold
    V_th_GrC   = -55 * mV
    V_th_GoC   = -50 * mV

    # Resting potential
    V_r_GrC    = -75 * mV
    V_r_GoC    = -55 * mV

    # Golgi cell reset potential
    V_reset_GoC = -55 * mV #-60

    # Synaptic weights
    fixed_w_e_GrC = 0.25 * nS #0.65 * nS
    w_i_GrC = 0.17 * nS

    fixed_w_e_GoC_M = 0.65 * nS #0.35  * nS
    w_e_GoC_GrC = 0.0 * nS

    # Stochastic fluctuating excitatory current
    sigma_n_GoC = 0.1 * nS
    sigma_n_GrC = 0.05 * nS

    tau_n   = 20 * ms

    #############
    # Equations #
    #############
    GrC_eqs = '''
    dv/dt   = (g_l*(E_l-v) + (g_e+g_n)*(E_e-v) + (g_i+g_t)*(E_i-v))/C_m : volt (unless refractory)
    dg_n/dt = (-g_n + sigma_n_GrC * sqrt(tau_n) * xi)/tau_n : siemens
    dg_e/dt = -g_e/tau_e : siemens
    dg_i/dt = -g_i/tau_i : siemens
    g_e_tot_GrC : siemens
    g_i_tot_GrC : siemens
    g_t = reduce_tonic * g_t_GrC : siemens
    reduce_tonic: 1
    '''

    GoC_eqs = '''
    dv/dt   = (g_l*(E_l-v) + (g_e+g_n)*(E_e-v))/C_m : volt (unless refractory)
    dg_e/dt = -g_e/tau_e : siemens
    dg_n/dt = (-g_n + sigma_n_GoC * sqrt(tau_n) * xi)/tau_n : siemens
    g_e_tot_GoC : siemens
    '''

    #####################
    # Mossy fiber input #
    #####################

    stim_times = [1010, 1020, 1030]            # Mossy fiber stimulation times
    nstim      = len(stim_times)
    mf_indices = np.arange(nmf)
    n_active   = round(nmf/10)           # Fraction of mfs active at each stimulation

    # Randomly select a subset of mossy fibers to spike at each stimulation
    # If you want a different subset of mossy fibers at each stimulation, move
    # the declaration of [active_indices] into the loop over stim_times

    random.seed(a=int(os.getenv("CHAT_SEED_TRIAL", 451)))
    active_indices = [mf_indices[i] for i in sorted(random.sample(range(len(mf_indices)),n_active))]
    indices    = []
    times      = []
    for j in range(nstim):
        indices.extend(active_indices)
        times.extend([stim_times[j]]*len(active_indices))
    times    = times * ms
    mf_input = SpikeGeneratorGroup(nmf, indices, times)


    ######################
    # Neuron populations #
    ######################
    if Condition == 'GABAzine':
        GrC = NeuronGroup(nGrC,
                          Equations(GrC_eqs,
                                    g_l   = g_l_GrC,
                                    E_l   = E_l_GrC,
                                    E_e   = E_e_GrC,
                                    E_i   = E_i_GrC,
                                    C_m   = C_m_GrC,
                                    tau_e = tau_e_decay_GrC,
                                    tau_i = tau_i_decay_GrC),
                          threshold  = 'v > V_th_GrC',
                          reset      = 'v = V_r_GrC',
                          refractory = 'tau_r_GrC',
                          method     = 'euler')
        GrC.v   = V_r_GrC
        GrC.reduce_tonic[:] = 0
    elif Condition == 'Control':
        GrC = NeuronGroup(nGrC,
                          Equations(GrC_eqs,
                                    g_l   = g_l_GrC,
                                    E_l   = E_l_GrC,
                                    E_e   = E_e_GrC,
                                    E_i   = E_i_GrC,
                                    C_m   = C_m_GrC,
                                    tau_e = tau_e_decay_GrC,
                                    tau_i = tau_i_decay_GrC),
                          threshold  = 'v > V_th_GrC',
                          reset      = 'v = V_r_GrC',
                          refractory = 'tau_r_GrC',
                          method     = 'euler')
        GrC.v   = V_r_GrC
        GrC.reduce_tonic[:] = 1
    elif Condition == 'ACh':
        GrC = NeuronGroup(nGrC,
                          Equations(GrC_eqs,
                                    g_l   = g_l_GrC,
                                    E_l   = E_l_GrC,
                                    E_e   = E_e_GrC,
                                    E_i   = E_i_GrC,
                                    C_m   = C_m_GrC,
                                    tau_e = tau_e_decay_GrC,
                                    tau_i = tau_i_decay_GrC),
                          threshold  = 'v > V_th_GrC',
                          reset      = 'v = V_r_GrC',
                          refractory = 'tau_r_GrC',
                          method     = 'euler')
        GrC.v   = V_r_GrC
        # Tonic reduction mean = 0.4 (60% reduction) variance = 0.07
        reduction_shape,reduction_scale = gamma_parms(0.4,0.07)
        GrC.reduce_tonic[:] = np.random.RandomState(seed=1).gamma(reduction_shape,reduction_scale,nGrC)
    else:
        print('ERROR: Unknown experimental condition')

    if Condition in ('Control','GABAzine'):
        GoC = NeuronGroup(nGoC,
                          Equations(GoC_eqs,
                                    g_l = g_l_GoC,
                                    E_l = E_l_GoC,
                                    E_e = E_e_GoC,
                                    E_i = E_i_GoC,
                                    C_m = C_m_GoC,
                                    tau_e = tau_e_decay_GoC),
                          threshold  = 'v > V_th_GoC',
                          reset      = 'v = V_r_GoC',
                          refractory = 'tau_r_GoC',
                          method     = 'euler')
    else:
        GoC = NeuronGroup(nGoC,
                          Equations(GoC_eqs,
                                    g_l = g_l_GoC,
                                    E_l = -55*mV,
                                    E_e = E_e_GoC,
                                    E_i = E_i_GoC,
                                    C_m = C_m_GoC,
                                    tau_e = tau_e_decay_GoC),
                          threshold  = 'v > V_th_GoC',
                          reset      = 'v = V_r_GoC',
                          refractory = 'tau_r_GoC',
                          method     = 'euler')
    GoC.v = V_r_GoC

    ###################
    # Connect neurons #
    ###################
    if Condition in ('Control', 'GABAzine'):
        # Mossy fiber onto GrCs
        GrC_M = Synapses(mf_input,GrC,
                         model  = '''w_e_GrC : siemens
                                     g_e_tot_GrC_post = g_e : siemens (summed)''',
                         on_pre = 'g_e += w_e_GrC')
        GrC_M.connect(p = conv_GrC_M/nmf)

        # Mossy fiber onto GoCs
        GoC_M = Synapses(mf_input,GoC,
                         model  = '''w_e_GoC_M : siemens
                                     g_e_tot_GoC_post = g_e : siemens (summed)''',
                         on_pre = 'g_e += w_e_GoC_M')
        GoC_M.connect(p = conv_GoC_M/nmf)
    elif Condition == 'ACh':
        # Mossy fiber onto GrCs
        GrC_M = Synapses(mf_input,GrC,
                         model  = '''w_e_GrC : siemens
                                     g_e_tot_GrC_post = g_e : siemens (summed)''',
                         on_pre = 'g_e += 0.4521*w_e_GrC')
        GrC_M.connect(p = conv_GrC_M/nmf)

        # Mossy fiber onto GoCs
        GoC_M = Synapses(mf_input,GoC,
                         model  = '''w_e_GoC_M : siemens
                                     g_e_tot_GoC_post = g_e: siemens (summed)''',
                         on_pre = 'g_e += 0.4578*w_e_GoC_M')
        GoC_M.connect(p = conv_GoC_M/nmf)

    # GoC onto GrC (inhibitory)
    if Condition == 'GABAzine':
        GrC_GoC = Synapses(GoC,GrC,
                           on_pre = 'g_i += 0 * nS',
                           delay  = tau_r_GrC)
    elif Condition in ('Control','ACh'):
        GrC_GoC = Synapses(GoC,GrC,
                           model  = 'g_i_tot_GrC_post = g_i : siemens (summed)',
                           on_pre = 'g_i += w_i_GrC',
                           delay  = tau_r_GrC)
    # seed(4321)
    GrC_GoC.connect(p = conv_GrC_GoC/nGoC)

    connections = pd.DataFrame({'MF': GrC_M.i[active_indices,:],'GrC': GrC_M.j[active_indices,:]})
    ffi_connections = pd.DataFrame({'GoC':GrC_GoC.i[:], 'GrC':GrC_GoC.j[:]})
    ################################
    # Randomize excitatory weights #
    ################################
    if random_weights:
        active_mf_GrC_weights = dict()
        active_mf_GoC_weights = dict()
        print('Randomizing synaptic weights')

        @network_operation(dt=runtime*ms, when='start')
        def update_input():
            weight_means = np.random.RandomState(seed=int(os.getenv("CHAT_SEED_WEIGHT_MEANS", 35))).gamma(weights_gamma_shape,weights_gamma_scale,n_active)
            # weight_sigma = 0.2
            tempidx = 0
            for i in active_indices:
                mu = weight_means[tempidx]
                # sigma = weight_sigma
                sigma = mu
                shape,scale = gamma_parms(mu,sigma)
                weight = np.random.RandomState(seed=trial_seed).gamma(shape,scale)*nS
                GrC_M.w_e_GrC[  i,:] = weight
                GoC_M.w_e_GoC_M[i,:] = weight
            # Fill in weight dictionary by mossy fiber number, add new mf key if non-existent
                if i in active_mf_GrC_weights.keys():
                    active_mf_GrC_weights[i].append(weight/nS)
                    active_mf_GoC_weights[i].append(weight/nS)
                else:
                    active_mf_GrC_weights[i] = [weight/nS]
                    active_mf_GoC_weights[i] = [weight/nS]
                tempidx += 1
    else:
        print('Using fixed synaptic weights')
        GrC_M.w_e_GrC[active_indices,:]   = fixed_w_e_GrC
        GoC_M.w_e_GoC_M[active_indices,:] = fixed_w_e_GoC_M



    ##############
    # Simulation #
    ##############

    # Set up granule cell spike monitor to collect spike trains
    M = SpikeMonitor(GrC)
    state_GrC  = StateMonitor(GrC, ['g_e_tot_GrC','g_i_tot_GrC'], dt=1*ms, record = True)
    # Store network state
    store(Condition)
    # Initialize matrix to store spike spike trains and conductance traces
    spikes = []
    cond_e_mats = []
    cond_i_mats = []
    # Number of trials to simulate
    ntrial = int(os.getenv("CHAT_NUM_TRIAL", 100))
    seeds = np.random.RandomState(seed=43).randint(1,1000,ntrial)
    for trial in range(ntrial):
        print('Trial {}'.format((trial+1)))
        # Restore network state
        restore(Condition)
        trial_seed = seeds[trial]
        # Run simulation and print progress report to console
        run(runtime * ms, report='stdout', report_period=1*second,
            profile=True)
        # Save spike times for each simulation
        spike_trains = M.spike_trains()
        spikes.append(spike_trains)
        # Save excitatory and inhibitory conductance traces for each simulation
        cond_e = state_GrC.g_e_tot_GrC/nS
        cond_i = state_GrC.g_i_tot_GrC/nS
        cond_e_mats.append(cond_e)
        cond_i_mats.append(cond_i)

    ############
    # Analysis #
    ############
    dt     = 1e-3
    ncol = int(runtime*1e-3/dt)
    mats   = []
    tuples = []
    for trial_idx  in range(ntrial):
        nrow    = len(spikes[trial_idx])
        arrays  = [np.full(nrow,trial_idx),range(nrow)]
        tuples += list(zip(*arrays))

        row  = np.concatenate([[key]*len(val) for key,val in spikes[trial_idx].items()])
        data = np.concatenate([val for key,val in spikes[trial_idx].items()])
        col  = data/dt
        trial_mat  = scipy.sparse.csr_matrix((data,(row.astype(int), col.astype(int))),shape = (nrow,ncol))
        mats.append(trial_mat)

    multi_index = pd.MultiIndex.from_tuples(tuples,names=['trial','neuron'])

    mat = scipy.sparse.vstack(mats)
    df = pd.DataFrame.sparse.from_spmatrix(mat, index=multi_index)

    mat_cond_e = np.vstack(cond_e_mats)
    df_cond_e = pd.DataFrame(mat_cond_e,index=multi_index)

    mat_cond_i = np.vstack(cond_i_mats)
    df_cond_i = pd.DataFrame(mat_cond_i,index=multi_index)

    spike_prob = np.zeros((nGrC,ncol))
    mean_cond_e = np.zeros((nGrC,mat_cond_e.shape[1]))
    mean_cond_i = np.zeros((nGrC,mat_cond_i.shape[1]))
    for neuron in range(nGrC):
        neuron_count = neuron+1
        if neuron_count < nGrC:
            print('{}: Compiling spike trains for neuron {} of {}'.format(Condition,neuron_count,nGrC), end='\r')
        else:
            print('{}: Compiling spike trains for neuron {} of {}'.format(Condition,neuron_count,nGrC), end='\n')
        st = df.loc[(slice(None),neuron),:].to_numpy()
        x = np.nan_to_num(st).astype(bool)
        p = np.mean(x,axis=0)
        spike_prob[neuron,:] = p

        this_cond_e = df_cond_e.loc[(slice(None),neuron),:].to_numpy()
        mean_this_cond_e = np.mean(this_cond_e,axis=0)
        mean_cond_e[neuron,:] = mean_this_cond_e

        this_cond_i = df_cond_i.loc[(slice(None),neuron),:].to_numpy()
        mean_this_cond_i = np.mean(this_cond_i,axis=0)
        mean_cond_i[neuron,:] = mean_this_cond_i

    mean_cond_e = mean_cond_e[np.unique(GrC_M.j[active_indices,:]),:]
    mean_cond_i = mean_cond_i[np.unique(GrC_M.j[active_indices,:]),:]

    if Condition == 'Control':
        if random_weights:
            np.save('randomized_controlSpikeProb',spike_prob)
            np.save('controlConnections',connections)
            np.save('controlWeights',active_mf_GrC_weights)
            np.save('controlCondE',mean_cond_e)
            np.save('controlCondI',mean_cond_i)
        else:
            np.save('controlSpikeProb',spike_prob)
    elif Condition == 'GABAzine':
        if random_weights:
            np.save('randomized_gabaSpikeProb',spike_prob)
            np.save('gabaConnections',connections)
            np.save('gabaWeights', active_mf_GrC_weights)
            np.save('gabaCondE',mean_cond_e)
            np.save('gabaCondI',mean_cond_i)
        else:
            np.save('gabaSpikeProb',spike_prob)
    elif Condition == 'ACh':
        if random_weights:
            np.save('randomized_achSpikeProb',spike_prob)
            np.save('achConnections',connections)
            np.save('achWeights', active_mf_GrC_weights)
            reduce_tonic = GrC.reduce_tonic[np.unique(GrC_M.j[active_indices,:])]
            np.save('tonicReduction',reduce_tonic)
            np.save('achCondE',mean_cond_e)
            np.save('achCondI',mean_cond_i)
        else:
            np.save('achSpikeProb',spike_prob)
    else:
        print('ERROR: Unknown experimental condition')

if __name__ == '__main__':
    main()
